### PR TITLE
Add dynamic fee tiers to payment stream contract

### LINKED ISSUE #504: [Contract] Implement Dynamic Fee Tier System based on Stream Volume
### Task Description
Support tiered protocol streaming fees in Soroban contracts to reduce fee percentages for high-volume ecosystem donors.

### Requirements & Scope
- **Target Area**: Soroban Rust Smart Contracts (`contracts/` / `src/contracts/`)
- **Key Objectives**:
  1. Implement complete, production-ready Rust code following Soroban SDK best practices.
  2. Include explicit authorization checks (`require_auth()`) and error handling enums.
  3. Write comprehensive unit tests in Rust covering success, failure, and edge cases.
  4. Ensure storage TTL management and efficient state footprint.

### Acceptance Criteria
- [ ] Code compiles cleanly with zero warnings (`cargo build --target wasm32-unknown-unknown`).
- [ ] All unit tests pass (`cargo test`).
- [ ] Proper error codes defined in `Error` enum.
- [ ] Complete documentation comments added to public functions.

### PR DESCRIPTION
**Pull request by opratem**
# Description

Implements a dynamic protocol fee tier system for the Soroban payment stream contracts. The protocol now supports configurable fee tiers based on stream volume, allowing high-volume contributors to benefit from reduced protocol fees while keeping fee calculations deterministic and efficient.

# Related Issue

Closes #504

# Type of Change

- [x] New feature
- [x] Documentation update

# Changes Made

- Implemented a configurable dynamic fee tier framework based on stream volume.
- Added deterministic fee calculation logic that automatically selects the appropriate fee tier.
- Added validation for fee tier configuration and volume thresholds.
- Extended the contract `Error` enum with typed errors for invalid fee configuration and calculation failures where applicable.
- Preserved efficient storage usage and applied proper Soroban storage TTL management.
- Added Rust documentation comments for all new public functions and updated existing documentation where necessary.

# Testing

- Added unit tests covering:
  - Fee calculation across multiple volume tiers.
  - Boundary conditions at fee tier thresholds.
  - Invalid fee tier configurations.
  - Authorization checks for fee administration.
  - Edge cases and error scenarios.
- Verified that existing contract functionality continues to work correctly.

```bash
cargo test
cargo build --target wasm32-unknown-unknown
```

All tests pass successfully with zero compiler warnings.

# Checklist

- [x] Code follows Soroban SDK best practices
- [x] Self-reviewed my code
- [x] Added comprehensive unit tests
- [x] Updated documentation (Rustdoc)
- [x] Proper authorization checks (`require_auth()`)
- [x] Efficient storage and TTL management
- [x] No new compiler warnings

# Screenshots

N/A — Smart contract changes only.

# Additional Notes

The dynamic fee tier system is designed to provide predictable and efficient fee calculations while rewarding higher-volume contributors with reduced protocol fees. The implementation integrates cleanly with the existing payment stream contracts without introducing breaking changes.

## Summary by CodeRabbit

* **New Features**
  * Added volume-based protocol fee tiers for payment streams.
  * Added tools to configure, view, and validate fee tiers.
  * Fees now adjust automatically based on a sender’s accumulated escrow volume.
  * Added safeguards for tier limits, ordering, and fee-rate boundaries.

* **Bug Fixes**
  * Improved contract error validation and handling for campaign operations.

* **Tests**
  * Added comprehensive coverage for fee tiers, volume tracking, fee calculation, authorization, and configuration updates.

### CODE REVIEW

File: contracts/campaign-funding/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,1501 @@
...
+        client.contribute(&contributor, &id, &9_999);
+        set_time(&env, 3_000);
+        client.trigger_expiry(&id);
+        client.claim_funds(&id);
+
+        // fee = 9_999 * 100 / 10_000 = 99 (integer division); net = 9_900.
+        assert_eq!(token_client.balance(&creator), 9_900);
+        assert_eq!(token_client.balance(&fee_collector), 99);
+    }
+}
```
[the remark this answers is not in the document]
Comment (opratem) [reply contracts/campaign-funding/src/lib.rs:1502]: done now

File: contracts/payment-stream/src/lib.rs [hunk 2]
```diff
@@ -221,6 +240,17 @@ impl PaymentStreamContract {
...
+        // Updated at stream creation so withdrawals automatically apply the
+        // correct tier rate without requiring additional state lookups.
+        let sender_vol_key = (Symbol::new(&env, "sv"), sender.clone());
+        let current_vol: i128 = env.storage().persistent()
+            .get(&sender_vol_key)
+            .unwrap_or(0);
+        let new_vol = current_vol.saturating_add(total_amount);
+        env.storage().persistent().set(&sender_vol_key, &new_vol);
+        env.storage().persistent().extend_ttl(&sender_vol_key, LEDGER_THRESHOLD, LEDGER_BUMP);
+
```
[the remark this answers is not in the document]
Comment (opratem) [reply contracts/payment-stream/src/lib.rs:713]: done now

File: contracts/payment-stream/src/lib.rs [hunk 3]
```diff
@@ -434,16 +464,56 @@ impl PaymentStreamContract {
...
+        let mut applicable_rate = general_rate;
+        let len = tiers.len();
+        for i in 0..len {
+            let tier = tiers.get(i).unwrap();
+            if volume >= tier.min_volume {
+                applicable_rate = tier.fee_rate;
+            }
+        }
+        applicable_rate
+    }
```
[the remark this answers is not in the document]
Comment (opratem) [reply contracts/payment-stream/src/lib.rs:1182]: done now

**Review by Utilitycoder — CHANGES_REQUESTED** [review]
Please fix up the merge conflict and update your PR. Kindly ensure you offramp with Fundable at [https://stellar.fundable.finance/offramp](https://stellar.fundable.finance/offramp)

**Review by Utilitycoder — CHANGES_REQUESTED** [review]
Please fix up the merge conflict and update your PR. Kindly ensure you offramp with Fundable at [https://stellar.fundable.finance/offramp](https://stellar.fundable.finance/offramp)

**Comment by Utilitycoder** [comment]
Closing this PR as it is older than a month. Kindly ensure you offramp with Fundable at [https://stellar.fundable.finance/offramp](https://stellar.fundable.finance/offramp)
